File: problems/24-knapsack-dp/001-partition-equal-subset-sum/python/solution.py
class Solution:
    def canPartition(self, nums: list[int]) -> bool:
        total = sum(nums)
        if total % 2:
            return False

        target = total // 2
        reachable = {0}

        for value in nums:
            reachable |= {current + value for current in reachable if current + value <= target}

        return target in reachable

File: problems/24-knapsack-dp/001-partition-equal-subset-sum/java/Solution.java
public class Solution {
    public boolean canPartition(int[] nums) {
        int total = 0;
        for (int value : nums) {
            total += value;
        }

        if (total % 2 != 0) {
            return false;
        }

        int target = total / 2;
        boolean[] dp = new boolean[target + 1];
        dp[0] = true;

        for (int value : nums) {
            for (int sum = target; sum >= value; sum--) {
                dp[sum] = dp[sum] || dp[sum - value];
            }
        }

        return dp[target];
    }
}
